WIP chore: Update readme (#842)
tensorchord/envd: README.md

<div align="center">
<h1>envd</h1>

> <p>Docker alternative for AI/ML applications</p>

Reviewer: I personally prefer `development environment` over `docker alternative`. 

Reasons:
- When searching about `docker alternative`, the results would by containerd/podman, which is not actually what we're referring to.
- Let's say if a researcher came to this project, he might think "I don't need docker, why even bother with having a docker alternative". The slogan itself should create an "Aha moment", especially for algorithm researchers/scientists who don't know about docker.

Reviewer 2: * I agree that `docker alternative` is ambiguous. It's more like `Dockerfile alternative`.
* I think we also support serving and we will add more serving-related features. So it's not just `development environment`
* We should pick up a slogan for both DS/ML and the infra team.

Reviewer 3: Second to that. We need to come up with a better one, ideally one that reflects what "envd" means and easy for people to remember. Are we settled with "envd" for the project name?

Author: I use Docker alternative from this tweet. https://twitter.com/Rami_Krispin/status/1540328916589498368

I think docker alternative may be easy to understand from the perspective of ds and algorithm engineers.

Reviewer: How about `Clean and portable workspaces`?

Author: I cannot get it from the text. The description is too general for me.

Author: > Second to that. We need to come up with a better one, ideally one that reflects what "envd" means and easy for people to remember. Are we settled with "envd" for the project name?

Yep, the name is envd.

Author: Let's discuss in https://github.com/tensorchord/envd/discussions/834

Author: I reverted the change here.

<a href='https://coveralls.io/github/tensorchord/envd?branch=main'><img src='https://coveralls.io/repos/github/tensorchord/envd/badge.svg?branch=main' alt='Coverage Status' /></a>
</p>

> **⚠️ envd is still under heavy development, and subject to change. it is not feature-complete or production-ready.**

envd (`ɪnˈvdɪ`) provides an alternative to Docker for AI/ML applications.

🐍 **Escape Dockerfile Hell** - Developing with Python, saving time on writing Dockerfiles, bash scripts, and Kubernetes YAML manifests

⏱️ **Save time** - Build the environment up to **7x faster**.

☁️ **Local & cloud** - `envd` images are OCI compatible, integrate with Docker and Kubernetes seamlessly.

🔁 **Repeatable builds & reproducible results** - You can reproduce the same environment on your laptop, public cloud VMs, or Docker containers, without any change in setup.

<p align="center">
  <img src="https://user-images.githubusercontent.com/5100735/187891206-b32160f1-5f65-415e-869a-d314763d6f4b.svg" width="700"/>
</p>

## Why use `envd`?

Environments built with `envd` provide the following features out-of-the-box:

> 🐍 **Life is short, use Python[^1]**

Reviewer 3: Did you want to keep these languages?

Author: Yep, but the build language is based on starlark, which is a dialect of python. Thus we may need to clarify this.

Reviewer 3: SGTM

<a href='https://coveralls.io/github/tensorchord/envd?branch=main'><img src='https://coveralls.io/repos/github/tensorchord/envd/badge.svg?branch=main' alt='Coverage Status' /></a>
</p>

> > **⚠️ envd is still under heavy development, and subject to change. it is not feature-complete or production-ready.**

Author: Thanks for the advice but I do not think we should emphasize this. The warning just shows that we are in the early stage.

![envd](https://user-images.githubusercontent.com/5100735/188821980-dcbd9069-b504-436a-9ffd-05ac5543a6d1.png)

[^1]: The build language is [starlark](https://docs.bazel.build/versions/main/skylark/language.html), which is a dialect of Python.

⏱️ **6x faster build**

`envd` adopts a multi-level cache mechanism to accelerate the building process. For example, the PyPI cache is shared across builds and thus the package will be cached if it has been downloaded before. It saves plenty of time, especially when you update the environment by trial and error.[^2]

<p align=center>
  <img src="https://user-images.githubusercontent.com/5100735/188601795-8c37f5a3-b13b-422b-816f-8a0c51f1f8b1.svg" width="65%"/>
</p>

[^2]: Docker without [buildkit](https://github.com/moby/buildkit)

☁️ **Local & cloud native**

<table>
<tr>
<td>Local development simplifies the debugging, but...</td> 
<td>Setup local & cloud native environment with envd</td>
</tr>
<tr>
<td>

❌ Can be complex to setup. When they break, you often need to run the whole setup.

> ❌ Resource intensive because of the constraints to your CPU, moemory and local GPU.

Reviewer: ```suggestion
❌ Intensive resource because of your CPU, memory, and local GPU constraints.
```

Author: `Intensive resource` looks weird to me. I am not sure if it works.

> ❌ Cannot be reproducible. Because not everyone has an identical setup.

Reviewer: ```suggestion
❌ Not reproducible. Because not everyone has an identical setup.
```

</td>
<td>

✅ AI infrastructure as code, reproduce the environment painlessly.

✅ Separate your environment to avoid impacting your local configuration.

> ✅ Use larger or more specialized hardware.

Reviewer: ```suggestion
✅ Use larger or more specialized hardwares.
```

Author: The noun hardware is uncountable, I think.
